diff --git a/src/http/ngx_http_core_module.c b/src/http/ngx_http_core_module.c
index a2ff53f82..dfc665979 100644
--- a/src/http/ngx_http_core_module.c
+++ b/src/http/ngx_http_core_module.c
@@ -2447,6 +2447,7 @@ ngx_http_subrequest(ngx_http_request_t *r,
     sr->http_version = r->http_version;
 
     sr->request_line = r->request_line;
+    sr->port = r->port;
     sr->uri = *uri;
 
     if (args) {
